File: LeetCode Solutions/October 2022 LeetCoding Challenge/Number of Dice Rolls With Target Sum.java
class Solution {
    public int numRollsToTarget(int n, int k, int target) {
        int MOD = 1000_000_000 + 7;
		int[][] dp = new int[n + 1][target + 1];
		dp[0][0] = 1;
		for (int i=1; i<=n; i++){
			for (int j=1; j<=target; j++){
				if (j > i * k){
					dp[i][j] = dp[i - 1][j - 1];
				} else {
					for (int l=1; l<=k && l<=j; l++){
						dp[i][j] = (dp[i][j] % MOD + dp[i - 1][j - l] % MOD) % MOD;
					}
				}
			}
		}
		return dp[n][target];
    }
}

File: LeetCode Solutions/October 2022 LeetCoding Challenge/Number of Dice Rolls With Target Sum.cpp
class Solution {
public:
    int numRollsToTarget(int n, int k, int target) {
        if (n > target || n * k < target) {
            return 0;
        }
        vector<vector<int>> dp(n+1, vector<int>(target+1, 0));
        int MOD = 1000000007;
        dp[0][0] = 1;
        for (int i=1; i<=n; i++) {
            for (int j=1; j<=target; j++) {
                dp[i][j] = (dp[i][j-1] + dp[i-1][j-1]) % MOD;
                if (j > k) {
                    dp[i][j] -= dp[i-1][j-k-1];
                    if (dp[i][j] < 0) {
                        dp[i][j] += MOD;
                    }
                }
            }
        }
        return dp[n][target];
    }
};